README.md
# mysql-monitor

#### Table of contents

1. [Overview] (#overview)
2. [Usage] (#usage)
3. [Notices] (#notices)
4. [Installation] (#installation)
5. [Development] (#development)

## Overview
This program scan and report all the TINYINT, SMALLINT, MEDIUMINT, INT, BIGINT columns where the highest value is too
close from the highest possible value allowed by the column type.

It handles UNSIGNED and SIGNED cases.

## Usage

```
usage: monitor.py [-h] [--username USERNAME] [--password [PASSWORD]]
                  [--host HOST] [--threshold THRESHOLD]
                  [--exclude EXCLUDE [EXCLUDE ...]] [--db DB [DB ...]]

optional arguments:
  -h, --help            show this help message and exit
  --username USERNAME, -u USERNAME
                        MySQL username
  --password [PASSWORD], -p [PASSWORD]
                        MySQL password
  --host HOST           MySQL host
  --threshold THRESHOLD, -t THRESHOLD
                        The alerting threshold (ex: 0.8 means alert when a
                        column max value is 80% of the max possible value
  --exclude EXCLUDE [EXCLUDE ...], -e EXCLUDE [EXCLUDE ...]
                        Database to exclude separated by a comma
  --db DB [DB ...], -d DB [DB ...]
                        Databases to analyse separated by a comma (default
                        all)
```

## Notices
BE CAREFUL:
 - It could be very slow (especially on heavy loaded servers or servers with a huge databases/tables count.
 You surely want to run this tool on a slave instead of a master

 - This script disable innodb_stats computing for optimizing performance_schema analysis and enable it at the end

 See: http://www.percona.com/blog/2011/12/23/solving-information_schema-slowness/

 If you interrupt this script (Ctrl+C ...) it is up to you to reactivate it using something like:
 set global innodb_stats_on_metadata=1;


## Installation

mysql-monitor is not yet released but when it will, it will be available on Pypi.

To install it with a [Python virtualenv](http://docs.python-guide.org/en/latest/dev/virtualenvs/) (recommended):

```
$> virtualenv myenv
$> source myenv/bin/activate
(myenv) $> pip install mysql-monitor
(myenv) $> mysql-monitor -h
```

Or on your system, without virtualenv:

```
$> pip install mysql-monitor
$> mysql-monitor -h
```

## Development

To start hacking on the project, run:

```
$> virtualenv myenv
$> source myenv/bin/activate
(myenv) $> pip install -e .
(myenv) $> mysql-monitor -h
```


CHANGES.md
Changelog
=========

0.1.0 (unreleased)
------------------

Initial version.


mysql_monitor/SchemaInformation.py
import atexit
from math import ceil

from MySQLdb import escape_string


class SchemaInformation(object):
    def __init__(self, db):
        self.excluded_db = []
        self.included_db = []
        self._db = db
        self._int_types = {
            'tinyint': 255,
            'smallint': 65535,
            'mediumint': 16777215,
            'int': 4294967295,
            'bigint': 18446744073709551615}

    def init_mysql_session(self):
        cursor = self._db.cursor()
        cursor.execute('SET GLOBAL innodb_stats_on_metadata=0')
        # Handle group concat limitation (see MySQL group_concat documentation)
        cursor.execute('SET SESSION group_concat_max_len = 1048576')
        atexit.register(self.enable_statistics)

    def enable_statistics(self):
        cursor = self._db.cursor()
        cursor.execute('SET GLOBAL innodb_stats_on_metadata=1')

    def get_columns_by_table(self):

        inc_db_stmt = ''
        if self.included_db:
            inc_db_stmt = 'AND TABLE_SCHEMA IN(%s)' % self.in_stmt(
                self.included_db)

        excl_db_stmt = ''
        if self.excluded_db:
            excl_db_stmt = 'AND TABLE_SCHEMA NOT IN(%s)' % self.in_stmt(
                self.excluded_db)

        sql = """
SELECT
  TABLE_SCHEMA,
  TABLE_NAME,
  GROUP_CONCAT(COLUMN_NAME) AS COLUMN_NAMES,
  GROUP_CONCAT(COLUMN_TYPE) AS COLUMN_TYPES
FROM information_schema.COLUMNS
WHERE 1
    %s
    %s
    AND DATA_TYPE IN (%s)
GROUP BY TABLE_SCHEMA, TABLE_NAME
ORDER BY TABLE_SCHEMA, TABLE_NAME, COLUMN_NAME
"""
        sql = sql % (
            inc_db_stmt, excl_db_stmt, self.in_stmt(self._int_types.keys())
        )

        cursor = self._db.cursor()
        cursor.execute(sql)

        return cursor.fetchall()

    def get_table_max_values(self, database, table, columns):
        cursor = self._db.cursor()

        max_expr = ', '.join(
            map(lambda x: 'MAX(`%s`) AS \'%s\'' % (x, x), columns)
        )
        sql = 'SELECT %s FROM %s.%s' % (max_expr, database, table)

        try:
            cursor.execute(sql)
        except Exception as error:
            print str(error)
            print sql
            exit(2)

        return cursor.fetchone()

    def in_stmt(self, l):
        return (', '.join(map(lambda x: "'" + escape_string(x) + "'", l)))

    def get_type_max_value(self, col_type, unsigned):
        if unsigned == 'unsigned':
            return self._int_types[col_type]
        else:
            return int(ceil(self._int_types[col_type] / 2))

    def _in_stmt(self, l):
        return (', '.join(map(lambda x: "'" + escape_string(x) + "'", l)))


mysql_overflows_monitor/__init__.py
from argparse import ArgumentParser
import atexit
import getpass
import re

from MySQLdb import connect
from MySQLdb.cursors import DictCursor

from .SchemaInformation import SchemaInformation


# TODO: Maybe monitor float types?
def monitor():
    # TODO: maybe add phpmyadmin here?
    excluded_db = ['mysql', 'information_schema', 'performance_schema']
    included_db = []

    arg_parser = ArgumentParser()
    arg_parser.add_argument('--username', '-u', default='root',
                            help='MySQL username')
    arg_parser.add_argument('--password', '-p', default='',
                            help='MySQL password', nargs='?')
    arg_parser.add_argument('--host', default='localhost',
                            help='MySQL host')
    arg_parser.add_argument('--threshold', '-t', default=0.8, type=float,
                            help="""The alerting threshold (ex: 0.8 means"""
                                 """ alert when a column max value is 80%%"""
                                 """ of the max possible value""")
    arg_parser.add_argument('--exclude', '-e', nargs='+', default=[],
                            help='Database to exclude separated by a comma')
    arg_parser.add_argument('--db', '-d', required=False, nargs='+',
                            help="""Databases to analyse separated by a"""
                                 """ comma (default all)""")

    args = arg_parser.parse_args()
    args.exclude += excluded_db

    password = args.password
    if args.password is None:
        password = getpass.getpass()

    try:
        # MySQL connection
        db = connect(host=args.host,
                     user=args.username,
                     passwd=password,
                     cursorclass=DictCursor)
    except Exception as error:
        print str(error)
        exit(2)

    atexit.register(db.close)

    # Configure schema analyser
    schema = SchemaInformation(db)

    # Handle database inc/exl parameters
    schema.excluded_db = args.exclude
    if args.db is not None:
        schema.included_db = args.db

    # Disabling InnoDB statistics for performances
    schema.init_mysql_session()

    print "Start"

    # Get column definitions
    columns = schema.get_columns_by_table()

    for definition in columns:
        # Get all max values for a given table

        columns_max_values = schema.get_table_max_values(
            definition['TABLE_SCHEMA'], definition['TABLE_NAME'],
            definition['COLUMN_NAMES'].split(','))

        table_cols = zip(definition['COLUMN_NAMES'].split(','),
                         definition['COLUMN_TYPES'].split(','))

        # Process column by column
        for name, full_type in table_cols:
            # Parsing column data to retrieve details, max values ...
            type, unsigned = re.split('\\s*\(\d+\)\s*', full_type)
            max_allowed = schema.get_type_max_value(type, unsigned)
            current_max_value = columns_max_values[name]

            # Calculate max values with threshold and comparing
            if (current_max_value >= int(max_allowed * args.threshold)):
                percent = round(
                    float(current_max_value) / float(max_allowed) * 100, 2)
                resting = max_allowed - current_max_value
                print ("""WARNING: (%s %s) %s.%s.%s max value is %s near"""
                       """(allowed=%s%%, resting=%s)""") % (
                          type, unsigned, definition['TABLE_SCHEMA'],
                          definition['TABLE_NAME'], name, current_max_value,
                          percent, resting)
    print "Done"


monitor.py
#!/usr/bin/env python

__author__ = "Jeremy Cohen Solal"

from argparse import ArgumentParser
import atexit, re, getpass

from MySQLdb import connect
from MySQLdb.cursors import DictCursor

from SchemaInformation import SchemaInformation


# TODO: Maybe monitor float types?
def main():
    # TODO: maybe add phpmyadmin here?
    excluded_db = ['mysql', 'information_schema', 'performance_schema']
    included_db = []

    arg_parser = ArgumentParser()
    arg_parser.add_argument('--username', '-u', default='root',
                            help='MySQL username')
    arg_parser.add_argument('--password', '-p', default='',
                            help='MySQL password', nargs='?')
    arg_parser.add_argument('--host', default='localhost',
                            help='MySQL host')
    arg_parser.add_argument('--threshold', '-t', default=0.8, type=float,
                            help="""The alerting threshold (ex: 0.8 means"""
                                 """ alert when a column max value is 80%%"""
                                 """ of the max possible value""")
    arg_parser.add_argument('--exclude', '-e', nargs='+', default=[],
                            help='Database to exclude separated by a comma')
    arg_parser.add_argument('--db', '-d', required=False, nargs='+',
                            help="""Databases to analyse separated by a"""
                                 """ comma (default all)""")

    args = arg_parser.parse_args()
    args.exclude += excluded_db

    password = args.password
    if (args.password is None) :
        password = getpass.getpass()

    try:
        # MySQL connection
        db = connect(host=args.host,
                     user=args.username,
                     passwd=password,
                     cursorclass=DictCursor)
    except Exception as error:
        print str(error)
        exit(2)

    atexit.register(db.close)

    # Configure schema analyser
    schema = SchemaInformation(db)

    # Handle database inc/exl parameters
    schema.excluded_db = args.exclude
    if args.db is not None:
        schema.included_db = args.db

    # Disabling InnoDB statistics for performances
    schema.init_mysql_session()

    print "Start"

    # Get column definitions
    columns = schema.get_columns_by_table()

    for definition in columns:
        # Get all max values for a given table

        columns_max_values = schema.get_table_max_values(
            definition['TABLE_SCHEMA'], definition['TABLE_NAME'],
            definition['COLUMN_NAMES'].split(','))

        table_cols = zip(definition['COLUMN_NAMES'].split(','),
                         definition['COLUMN_TYPES'].split(','))

        # Process column by column
        for name, full_type in table_cols:
            # Parsing column data to retrieve details, max values ...
            type, unsigned = re.split('\\s*\(\d+\)\s*', full_type)
            max_allowed = schema.get_type_max_value(type, unsigned)
            current_max_value = columns_max_values[name]

            # Calculate max values with threshold and comparing
            if (current_max_value >= int(max_allowed * args.threshold)):
                percent = round(
                    float(current_max_value) / float(max_allowed) * 100, 2)
                resting = max_allowed - current_max_value
                print ("""WARNING: (%s %s) %s.%s.%s max value is %s near"""
                       """(allowed=%s%%, resting=%s)""") % (
                          type, unsigned, definition['TABLE_SCHEMA'],
                          definition['TABLE_NAME'], name, current_max_value,
                          percent, resting)
    print "Done"

if __name__ == '__main__':
    main()
